File: FlowCyPy/opto_electronics/detector.py
from typing import Optional

import numpy as np
from TypedUnit import (
    Angle,
    AnyUnit,
    Current,
    Frequency,
    Length,
    Responsitivity,
    ureg,
    validate_units,
)

from FlowCyPy.signal_generator import SignalGenerator
from FlowCyPy.physical_constant import PhysicalConstant
from FlowCyPy.utils import dataclass, config_dict, StrictDataclassMixing


@dataclass(config=config_dict, kw_only=True, unsafe_hash=True)
class Detector(StrictDataclassMixing):
    """
    Represents a photodetector for flow cytometry simulations.

    This class simulates a photodetector that captures light scattering signals in a flow
    cytometry setup. It models the detector response by incorporating detector specific
    transformations and optional noise sources into the signal processing workflow.

    The detector supports bandwidth dependent operations such as dark current noise and
    shot noise modeling. The bandwidth can be provided explicitly to each method, or stored
    directly on the detector instance through the ``bandwidth`` attribute. If neither is
    defined, the corresponding bandwidth dependent computation is skipped.

    Parameters
    ----------
    name : str, optional
        A unique identifier for the detector. If not provided, a unique ID is generated.
    phi_angle : Angle
        The primary azimuthal angle of incidence for the detector.
    numerical_aperture : float
        The numerical aperture of the detector.
    cache_numerical_aperture : float, optional
        The numerical aperture of the caching element placed in front of the detector.
        Default is 0.
    responsivity : Current / Power, optional
        Detector responsivity in ampere per watt. Default is 1 A/W.
    dark_current : Current, optional
        Dark current in ampere. Default is 0 A.
    bandwidth : Frequency, optional
        Detector bandwidth. If ``None``, bandwidth dependent effects are not applied
        unless a bandwidth is explicitly provided to the relevant methods.
    gamma_angle : Angle, optional
        The complementary longitudinal angle of incidence. Default is 0 degree.
    sampling : int, optional
        Number of spatial sampling points used to define the detector support.
        Default is 200.
    channel : str, optional
        Signal channel associated with this detector. Default is ``"scattering"``.
    """

    phi_angle: Angle
    numerical_aperture: float

    cache_numerical_aperture: float = 0
    gamma_angle: Optional[Angle] = 0 * ureg.degree
    sampling: Optional[int] = 200
    responsivity: Optional[Responsitivity] = 1 * ureg.ampere / ureg.watt
    dark_current: Optional[Current] = 0.0 * ureg.ampere
    bandwidth: Optional[Frequency] = None
    name: Optional[str] = None
    channel: str = "scattering"

    def __post_init__(self) -> None:
        """
        Finalize detector initialization.

        If no explicit name is provided, a unique identifier derived from the object
        instance is assigned.
        """
        if self.name is None:
            self.name = str(id(self))

    def _resolve_bandwidth(
        self,
        bandwidth: Optional[Frequency],
    ) -> Optional[Frequency]:
        """
        Return the effective bandwidth to use for a computation.

        The explicitly provided bandwidth takes precedence. If it is ``None``,
        the detector's own ``bandwidth`` attribute is used. If both are ``None``,
        the computation should be skipped by the caller.

        Parameters
        ----------
        bandwidth : Frequency or None
            Bandwidth provided directly to a method.

        Returns
        -------
        Frequency or None
            Effective bandwidth to use.
        """
        if bandwidth is not None:
            return bandwidth

        return self.bandwidth

    def apply_dark_current_noise(
        self,
        signal: Current,
        bandwidth: Optional[Frequency] = None,
    ) -> Current:
        r"""
        Apply dark current noise directly to a detector current signal.

        Dark current noise is bandwidth dependent. The method first resolves the
        effective bandwidth from the method argument or from ``self.bandwidth``.
        If no bandwidth is available, the input signal is returned unchanged.

        The dark current noise standard deviation is computed as:

        .. math::

            \sigma_{\mathrm{dark}} = \sqrt{2 q I_d B}

        where:

        - :math:`q` is the elementary charge
        - :math:`I_d` is the dark current
        - :math:`B` is the bandwidth

        A Gaussian random term with mean equal to the detector dark current and
        standard deviation :math:`\sigma_{\mathrm{dark}}` is added directly to
        each sample of the signal.

        Parameters
        ----------
        signal : Current
            Input detector current signal.
        bandwidth : Frequency or None, optional
            Signal bandwidth. If ``None``, ``self.bandwidth`` is used. If both are
            undefined, dark current noise is skipped.

        Returns
        -------
        Current
            Signal with dark current noise added.
        """
        bandwidth = self._resolve_bandwidth(bandwidth)

        if bandwidth is None:
            return signal

        standard_deviation_noise = np.sqrt(
            2 * 1.602176634e-19 * ureg.coulomb * self.dark_current * bandwidth
        )

        signal_magnitude = np.asarray(signal.to("ampere").magnitude, dtype=float)

        if signal_magnitude.size == 0:
            raise ValueError("Signal array is empty.")

        noisy_signal_magnitude = signal_magnitude + np.random.normal(
            loc=self.dark_current.to("ampere").magnitude,
            scale=standard_deviation_noise.to("ampere").magnitude,
            size=signal_magnitude.shape,
        )

        return noisy_signal_magnitude * ureg.ampere

    def _get_optical_power_to_photon_factor(
        self,
        wavelength: Length,
        bandwidth: Frequency,
    ) -> AnyUnit:
        """
        Compute the conversion factor from optical power to photon count.

        This factor represents the mean number of photons collected during one
        effective sampling interval associated with the provided bandwidth.

        Parameters
        ----------
        wavelength : Length
            Optical wavelength of the incident light.
        bandwidth : Frequency
            Signal bandwidth.

        Returns
        -------
        AnyUnit
            Conversion factor in ``1 / watt``.
        """
        energy_per_photon = PhysicalConstant.h * PhysicalConstant.c / wavelength
        sampling_interval = 1 / (bandwidth * 2)

        optical_power_to_photon_count_conversion_factor = (
            sampling_interval / energy_per_photon
        ).to("1 / watt")

        return optical_power_to_photon_count_conversion_factor

File: FlowCyPy/cpp/opto_electronics/detector/detector.cpp
#include "detector.h"

#include <random>
#include <sstream>
#include <cstdint>

#include <utils/constants.h>


Detector::Detector(
    const double phi_angle,
    const double numerical_aperture,
    const double cache_numerical_aperture,
    const double gamma_angle,
    const int sampling,
    const double responsivity,
    const double dark_current,
    const double bandwidth,
    const std::string& name
)
    : phi_angle(phi_angle),
      numerical_aperture(numerical_aperture),
      cache_numerical_aperture(cache_numerical_aperture),
      gamma_angle(gamma_angle),
      sampling(sampling),
      responsivity(responsivity),
      dark_current(dark_current),
      bandwidth(bandwidth),
      name(name)
{
    if (std::isnan(this->phi_angle)) {
        throw std::invalid_argument("Detector phi_angle must be defined.");
    }

    if (std::isnan(this->gamma_angle)) {
        throw std::invalid_argument("Detector gamma_angle must be defined.");
    }

    if (std::isnan(this->numerical_aperture) || this->numerical_aperture < 0.0) {
        throw std::invalid_argument("Detector numerical_aperture must be non negative.");
    }

    if (
        std::isnan(this->cache_numerical_aperture) ||
        this->cache_numerical_aperture < 0.0
    ) {
        throw std::invalid_argument(
            "Detector cache_numerical_aperture must be non negative."
        );
    }

    if (this->sampling <= 0) {
        throw std::invalid_argument("Detector sampling must be strictly positive.");
    }

    if (std::isnan(this->responsivity) || this->responsivity < 0.0) {
        throw std::invalid_argument("Detector responsivity must be non negative.");
    }

    if (std::isnan(this->dark_current) || this->dark_current < 0.0) {
        throw std::invalid_argument("Detector dark_current must be non negative.");
    }

    if (!std::isnan(this->bandwidth) && this->bandwidth <= 0.0) {
        throw std::invalid_argument(
            "Detector bandwidth must be strictly positive when provided."
        );
    }

    if (this->name.empty()) {
        this->name = std::to_string(
            reinterpret_cast<std::uintptr_t>(this)
        );
    }
}


bool Detector::has_bandwidth() const {
    return !std::isnan(this->bandwidth);
}


void Detector::clear_bandwidth() {
    this->bandwidth = std::numeric_limits<double>::quiet_NaN();
}


bool Detector::bandwidth_is_defined(const double bandwidth) const {
    return !std::isnan(bandwidth);
}


double Detector::resolve_bandwidth(
    const double bandwidth
) const {
    if (!std::isnan(bandwidth)) {
        if (bandwidth <= 0.0) {
            throw std::invalid_argument(
                "Detector method bandwidth must be strictly positive when provided."
            );
        }

        return bandwidth;
    }

    return this->bandwidth;
}


std::vector<double> Detector::apply_dark_current_noise(
    const std::vector<double>& signal,
    const double bandwidth
) const {
    const double effective_bandwidth = this->resolve_bandwidth(bandwidth);

    if (!this->bandwidth_is_defined(effective_bandwidth)) {
        return signal;
    }

    if (signal.empty()) {
        throw std::invalid_argument("Signal array is empty.");
    }

    const double standard_deviation_noise = std::sqrt(
        2.0 *
        Constants::e *
        this->dark_current *
        effective_bandwidth
    );

    std::random_device random_device;
    std::mt19937_64 random_number_generator(random_device());
    std::normal_distribution<double> normal_distribution(
        this->dark_current,
        standard_deviation_noise
    );

    std::vector<double> noisy_signal(signal.size());

    for (size_t index = 0; index < signal.size(); ++index) {
        noisy_signal[index] = signal[index] + normal_distribution(random_number_generator);
    }

    return noisy_signal;
}


double Detector::get_optical_power_to_photon_factor(
    const double wavelength,
    const double bandwidth
) const {
    if (std::isnan(wavelength) || wavelength <= 0.0) {
        throw std::invalid_argument("Detector wavelength must be strictly positive.");
    }

    if (std::isnan(bandwidth) || bandwidth <= 0.0) {
        throw std::invalid_argument("Detector bandwidth must be strictly positive.");
    }

    const double energy_per_photon = Constants::plank * Constants::light_speed / wavelength;
    const double sampling_interval = 1.0 / (bandwidth * 2.0);

    return sampling_interval / energy_per_photon;
}


std::string Detector::repr() const {
    return
        "Detector("
        "name='" + this->name + "'" +
        ", phi_angle=" + std::to_string(this->phi_angle) + "rad" +
        ", numerical_aperture=" + std::to_string(this->numerical_aperture) +
        ", cache_numerical_aperture=" + std::to_string(this->cache_numerical_aperture) +
        ", gamma_angle=" + std::to_string(this->gamma_angle) + "rad" +
        ", sampling=" + std::to_string(this->sampling) +
        ", responsivity=" + std::to_string(this->responsivity) + "A/W" +
        ", dark_current=" + std::to_string(this->dark_current) + "A" +
        ", bandwidth=" + (
            std::isnan(this->bandwidth) ? std::string("None") : std::to_string(this->bandwidth)
        ) + "Hz"
        ")";
}
